File: dotnet/src/Envdoctor/Commands/DiffCommand.cs
using Envdoctor.Core;
using Envdoctor.Detectors;
using Envdoctor.Utils;

namespace Envdoctor.Commands;

public sealed class DiffArgs
{
    public string? Root { get; set; }
    public string EnvA { get; set; } = "";
    public string EnvB { get; set; } = "";
    public bool Json { get; set; }
}

public static class DiffCommand
{
    /// `envdoctor diff <env1> <env2>` — compare variable sets across environments.
    public static int Run(DiffArgs args)
    {
        var root = Path.GetFullPath(args.Root ?? ".");

        var (model, _) = Pipeline.LoadProject(root);
        var labelA = Shared.NormalizeEnvLabel(args.EnvA);
        var labelB = Shared.NormalizeEnvLabel(args.EnvB);

        var available = new SortedSet<string>(StringComparer.Ordinal);
        foreach (var f in model.EnvFiles)
        {
            if (f.Environment is not null && f.Environment != "example")
                available.Add(f.Environment);
        }

        Shared.ReportParseErrors(model, root);

        if (!available.Contains(labelA) || !available.Contains(labelB))
        {
            if (!available.Contains(labelA))
                Console.Error.WriteLine($"error Environment \"{labelA}\" has no files in this project.");
            if (!available.Contains(labelB))
                Console.Error.WriteLine($"error Environment \"{labelB}\" has no files in this project.");
            Console.Error.WriteLine($"  Available: {(available.Count == 0 ? "none" : string.Join(", ", available))}");
            return 2;
        }

        var entries = EnvironmentDiffDetector.CompareEnvironments(model, labelA, labelB);
        var missingCount = entries.Count(e => !e.PresentInBoth);

        if (args.Json)
        {
            var variables = entries.Select(e => (object)new JsonObject
            {
                { "name", e.Name },
                { "status", e.PresentInBoth ? "same" : "missing" },
                { "missingIn", e.PresentInBoth ? null : e.PresentInA ? labelB : labelA },
            }).ToList();
            Console.Out.WriteLine(Json.Pretty(new JsonObject
            {
                { "environments", new List<object?> { labelA, labelB } },
                { "exitCode", missingCount > 0 ? 1 : 0 },
                { "total", entries.Count },
                { "missing", missingCount },
                { "variables", variables },
            }));
            return missingCount > 0 ? 1 : 0;
        }

        const string title = "ENVIRONMENT DIFF";
        Console.Out.WriteLine(title);
        Console.Out.WriteLine(new string('─', title.Length * 2) + "\n");
        Console.Out.WriteLine($"  {labelA} → {labelB}\n");

        foreach (var entry in entries)
        {
            if (entry.PresentInBoth)
                Console.Out.WriteLine($"  ✓ {entry.Name}  present in both");
            else if (entry.PresentInA)
                Console.Out.WriteLine($"  ❌ {entry.Name}  missing in {labelB}");
            else
                Console.Out.WriteLine($"  ❌ {entry.Name}  missing in {labelA}");
        }
        Console.Out.WriteLine(
            $"\n  Summary: {entries.Count} variables · {missingCount} missing · {entries.Count - missingCount} present in both");

        return missingCount > 0 ? 1 : 0;
    }
}

File: swift/Sources/EnvdoctorKit/Commands/DiffCommand.swift
import Foundation

public struct DiffArgs {
    public var root: String?
    public var envA: String
    public var envB: String
    public var json = false

    public init(envA: String, envB: String) {
        self.envA = envA
        self.envB = envB
    }
}

public enum DiffCommand {
    /// `envdoctor diff <env1> <env2>` — compare variable sets across environments.
    public static func run(_ args: DiffArgs) throws -> Int {
        let root = resolveRoot(args.root)
        let context = try Pipeline.loadProject(root)
        let labelA = normalizeEnvLabel(args.envA)
        let labelB = normalizeEnvLabel(args.envB)

        var availableSet = Set<String>()
        var availableOrder: [String] = []
        for f in context.model.envFiles {
            if let env = f.environment, !env.isEmpty, env != "example", !availableSet.contains(env) {
                availableSet.insert(env)
                availableOrder.append(env)
            }
        }
        let available = availableOrder

        reportParseErrors(context.model, root)

        if !available.contains(labelA) || !available.contains(labelB) {
            if !available.contains(labelA) {
                FileHandle.standardError.write("error Environment \"\(labelA)\" has no files in this project.\n")
            }
            if !available.contains(labelB) {
                FileHandle.standardError.write("error Environment \"\(labelB)\" has no files in this project.\n")
            }
            FileHandle.standardError.write("  Available: \(available.isEmpty ? "none" : available.joined(separator: ", "))\n")
            return 2
        }

        let entries = compareEnvironments(context.model, labelA, labelB)
        let missingCount = entries.filter { !$0.presentInBoth }.count

        if args.json {
            var variables: [JSONValue] = []
            for e in entries {
                var variable = JSONObject()
                variable.set("name", .string(e.name))
                variable.set("status", .string(e.presentInBoth ? "same" : "missing"))
                if e.presentInBoth {
                    variable.set("missingIn", .null)
                } else {
                    variable.set("missingIn", .string(e.presentInA ? labelB : labelA))
                }
                variables.append(.object(variable))
            }
            var rootJson = JSONObject()
            rootJson.set("environments", .array([.string(labelA), .string(labelB)]))
            rootJson.set("exitCode", .int(missingCount > 0 ? 1 : 0))
            rootJson.set("total", .int(entries.count))
            rootJson.set("missing", .int(missingCount))
            rootJson.set("variables", .array(variables))
            print(Json.pretty(.object(rootJson)))
            return missingCount > 0 ? 1 : 0
        }

        let title = "ENVIRONMENT DIFF"
        print(title)
        print(rule(title.count * 2) + "\n")
        print("  \(labelA) → \(labelB)\n")

        for entry in entries {
            if entry.presentInBoth {
                print("  ✓ \(entry.name)  present in both")
            } else if entry.presentInA {
                print("  ❌ \(entry.name)  missing in \(labelB)")
            } else {
                print("  ❌ \(entry.name)  missing in \(labelA)")
            }
        }
        print("\n  Summary: \(entries.count) variables · \(missingCount) missing · \(entries.count - missingCount) present in both")

        return missingCount > 0 ? 1 : 0
    }
}
